File: ThefieldofdreamsClient/src/app/models/campaign.ts
export type CampaignType = 'QRBased' | 'PredictionBased' | 'ReferralBased' | 'TourismBased' | 'Hybrid';
export type CampaignStatus = 'Draft' | 'Scheduled' | 'Active' | 'Paused' | 'Completed' | 'Cancelled';
export type CampaignRuleType =
  | 'RequireRole'      // {"role":"passenger"}
  | 'MinimumPoints'    // {"points":100}
  | 'MaxScansPerUser'  // {"max":3}
  | 'TargetUserGroup'  // {"group":"VIP"}
  | 'TimeWindow'       // {"startHour":9,"endHour":18}
  | 'GeoZone'          // {"wkt":"POLYGON((lon lat,...))"}
  | 'PortArrival';     // {"portLocationId":"guid"}

export interface CampaignRule {
  id: string;
  campaignId: string;
  ruleType: CampaignRuleType;
  parametersJson: string;
}

export interface Campaign {
  id: string;
  name: string;
  description?: string;
  sponsorId: string;
  type: CampaignType;
  startDate: string;
  endDate: string;
  budget?: number;
  totalImpressions: number;
  totalClicks: number;
  totalConversions: number;
  status: CampaignStatus;
  priority: number;
}

export interface CreateCampaignRequest {
  name: string;
  description?: string;
  sponsorId: string;
  type: CampaignType;
  startDate: string;
  endDate: string;
  budget?: number;
  status?: CampaignStatus;
  priority?: number;
}

export interface UpdateCampaignRequest {
  name: string;
  description?: string;
  type: CampaignType;
  startDate: string;
  endDate: string;
  budget?: number;
  status: CampaignStatus;
  priority?: number;
}

export interface QRCampaign {
  id: string;
  campaignName: string;
  description?: string;
  campaignId: string;
  qrCodeId: string;
  maxScansPerUser: number;
  totalScanLimit: number;
  currentScanCount: number;
  basePointsReward: number;
  bonusPointsForFirstScan?: number;
  rewardCode?: string;
  discountPercentage?: string;
  startDate?: string;
  endDate?: string;
  isActive: boolean;
  targetUserGroup?: string;
  minimumUserLevel?: number;
  uniqueUsersScanned: number;
  conversionCount: number;
  conversionRate?: number;
  successMessage?: string;
  redirectUrl?: string;
  sponsorOfferCode?: string;
  isSponsorExclusive: boolean;
}

export interface CreateQRCampaignRequest {
  campaignName: string;
  description?: string;
  campaignId: string;
  qrCodeId: string;
  maxScansPerUser: number;
  totalScanLimit: number;
  basePointsReward: number;
  bonusPointsForFirstScan?: number;
  rewardCode?: string;
  discountPercentage?: string;
  startDate?: string;
  endDate?: string;
  isActive: boolean;
  targetUserGroup?: string;
  minimumUserLevel?: number;
  successMessage?: string;
  redirectUrl?: string;
  sponsorOfferCode?: string;
  isSponsorExclusive: boolean;
}

export interface UpdateQRCampaignRequest extends CreateQRCampaignRequest {}

File: Thefieldofdreams.Domain/Entities/Campaign.cs
namespace Thefieldofdreams.Domain.Entities
{
    
    public class Campaign : BaseEntity
    {
        public required string Name { get; set; }
        public string? Description { get; set; }
        public Guid SponsorId { get; set; }
        public CampaignType Type { get; set; }
        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }
        public decimal? Budget { get; set; }
        public int TotalImpressions { get; set; } = 0;
        public int TotalClicks { get; set; } = 0;
        public int TotalConversions { get; set; } = 0;
        public CampaignStatus Status { get; set; } = CampaignStatus.Draft;
        public int Priority { get; set; } = 0;

        public virtual Sponsor Sponsor { get; set; }
        public virtual ICollection<QRCampaign> QRCampaigns { get; set; } = new List<QRCampaign>();
        public virtual ICollection<CampaignRule> Rules { get; set; } = new List<CampaignRule>();
    }

    public enum CampaignType
    {
        QRBased,
        PredictionBased,
        ReferralBased,
        TourismBased,
        Hybrid
    }

    public enum CampaignStatus
    {
        Draft,
        Scheduled,
        Active,
        Paused,
        Completed,
        Cancelled
    }
}